[RPC] Ensure _wait_all_workers doesn't swallow exception.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61094**

`_wait_all_workers` was swallowing exceptions and as a result if there
were any errors it would still continue with rpc_agent.join() which would hang
since something already failed before.

To fix this, I've ensured that wait_all_workers throws and in that case we just
proceed with an ungraceful shutdown without joining.

Differential Revision: [D29509286](https://our.internmc.facebook.com/intern/diff/D29509286/)

pritamdamania committed:
[RPC] Ensure _wait_all_workers doesn't swallow exception.

`_wait_all_workers` was swallowing exceptions and as a result if there
were any errors it would still continue with rpc_agent.join() which would hang
since something already failed before.

To fix this, I've ensured that wait_all_workers throws and in that case we just
proceed with an ungraceful shutdown without joining.

Differential Revision: [D29509286](https://our.internmc.facebook.com/intern/diff/D29509286/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4056,6 +4056,47 @@ def test_user_rref_backward(self):
+                rpc.api._broadcast_to_followers = og_func
+                rpc.api._delete_all_user_and_unforked_owner_rrefs = og_rref_func
+        else:
+            rpc.shutdown()
```
the purpose of the test is to make sure this call doesn't hang right? Is there anything we can assert here?

pritamdamania87 replied:
Right, it ensures this call doesn't hang. I'm not sure if we actually need to assert something here since we are just looking for successful shutdown.

rohan-varma reviewed (approved):
LGTM, thanks for improving the shutdown error handling! Also looks like a couple of the RPC tests are failing in CI.

pritamdamania committed:
Update on "[RPC] Ensure _wait_all_workers doesn't swallow exception."


`_wait_all_workers` was swallowing exceptions and as a result if there
were any errors it would still continue with rpc_agent.join() which would hang
since something already failed before.

To fix this, I've ensured that wait_all_workers throws and in that case we just
proceed with an ungraceful shutdown without joining.

Differential Revision: [D29509286](https://our.internmc.facebook.com/intern/diff/D29509286/)

[ghstack-poisoned]

lw commented on torch/distributed/rpc/api.py:
```diff
@@ -331,9 +332,20 @@ def shutdown(graceful=True):
     """
     if graceful:
-        _wait_all_workers()
+        try:
+            _wait_all_workers()
+        except RuntimeError as ex:
+            # Following steps are bound to fail, so just finalize
+            # shutdown and exit.
+            _finalize_shutdown()
+            return
+
         _delete_all_user_and_unforked_owner_rrefs()
         _get_current_rpc_agent().join(shutdown=True)
+
+    _finalize_shutdown()
+
+def _finalize_shutdown():
```
Nit: this seems a nice use-case for the little know `try-else` syntax:
```suggestion
        try:
            _wait_all_workers()
        else:
            _delete_all_user_and_unforked_owner_rrefs()
            _get_current_rpc_agent().join(shutdown=True)
```

pritamdamania87 replied:
My understanding is that else clause will run only if there are no exceptions. But in case there is an exception, we still need to catch it and call `finalize_shutdown` right? Otherwise the exception from `_wait_all_workers` will just be propagated up the stack without shutting down the RPC framework.

lw replied:
Can't this be achieved with a try-else-finally clause?

pritamdamania87 replied:
I guess try-finally should be sufficient?

lw replied:
Yeah probably. Possibly try-else-finally could be "clearer" as it shows which of the three statements we expect could raise an exception. Up to you.

pritamdamania committed:
Update on "[RPC] Ensure _wait_all_workers doesn't swallow exception."


`_wait_all_workers` was swallowing exceptions and as a result if there
were any errors it would still continue with rpc_agent.join() which would hang
since something already failed before.

To fix this, I've ensured that wait_all_workers throws and in that case we just
proceed with an ungraceful shutdown without joining.

Differential Revision: [D29509286](https://our.internmc.facebook.com/intern/diff/D29509286/)

[ghstack-poisoned]

pritamdamania committed:
Update on "[RPC] Ensure _wait_all_workers doesn't swallow exception."


`_wait_all_workers` was swallowing exceptions and as a result if there
were any errors it would still continue with rpc_agent.join() which would hang
since something already failed before.

To fix this, I've ensured that wait_all_workers throws and in that case we just
proceed with an ungraceful shutdown without joining.

Differential Revision: [D29509286](https://our.internmc.facebook.com/intern/diff/D29509286/)

[ghstack-poisoned]
